_posts/2019-02-15-introducing-aquery.md

```diff
@@ -0,0 +1,116 @@
+
+`aquery` is already being used in our various workflows. An example usage of `aquery` can be found in the Bazel issue [#6861](https://github.com/bazelbuild/bazel/issues/6861), where we are migrating legacy CROSSTOOL fields. In this case, Bazel users would run a migration tool, and then use `aquery` to verify that the migration tool works properly, in particular:
+
+- The actions generated while building the same target before & after running the migration tool are the same.
+- The command lines run for each action are the same.
+
+The specific usage is implemented in [`aquery_differ` tool](https://github.com/bazelbuild/bazel/blob/master/tools/aquery_differ/aquery_differ.py). This also serves as an example of how tools can be built on top of `aquery`.
+
+## Background & Motivation
+
+Apart from providing the ability to build & test your projects, Bazel also offers insights into how those processes happen with [`query`](https://docs.bazel.build/versions/master/query-how-to.html) and [`cquery`](https://docs.bazel.build/versions/master/cquery.html). These existing tools have been very helpful with answering the questions about dependencies of targets in your Bazel project.
+
+The Bazel build process consists of 3 phases<sup>[1](#footnote1)</sup>: Loading, Analysis and Execution. `query` operates on the post-loading phase target graph, which makes it unaware of the configurations of these targets. `cquery` moves it further down the building process and queries the post-analysis Configured Targets, thus includes the actual configurations.
+
+The topology of the Configured Target Graph closely resembles the dependency graph of targets established by the BUILD files. It offers information on the dependency between targets in a build, but not on the actual build actions that will be run to execute that build. To gain insights on the exact actions executed in a build, we have to go one level deeper, to the Action Graph.
+
+Enter `aquery`.
+
+![bazel queries and phases](/assets/bazel-queries.png)
+
+`aquery` runs on the Configured Target Graph and queries the Action Graph. The Action Graph<sup>[2](#footnote2)</sup> is the result of the Analysis Phase. It is a bipartite graph with the following types of nodes:
+
+- Artifacts: either a source file or any output file produced by an action
+- Actions: the functional step that takes a list of Artifacts as input and outputs a list of Artifacts. Note that any (output) artifact is produced by exactly one action. The Action Graph conveys explicit step-by-step instructions on how the build would be executed.
+
+With `aquery`, it is now possible to tap into that knowledge.
+
+## `bazel aquery`
+
+`aquery` is useful when we are interested in the properties of the Actions/Artifacts in the Action Graph. The basic structure of `aquery` output is as follows:
+
```

Comment:
I might clarify somewhere that the query language is the same as for query and cquery?

Reply:
Done.